quantum/pointing_device/pointing_device_drivers.c

```diff
@@ -268,24 +269,64 @@ mouse_xy_report_t pimoroni_trackball_adapt_values(clamp_range_t* offset) {
 }
 
 report_mouse_t pimoroni_trackball_get_report(report_mouse_t mouse_report) {
-    static uint16_t      debounce      = 0;
-    static uint8_t       error_count   = 0;
-    pimoroni_data_t      pimoroni_data = {0};
-    static clamp_range_t x_offset = 0, y_offset = 0;
+    static uint16_t debounce    = 0;
+    static uint8_t  error_count = 0;
+
+    static fast_timer_t began_motion, last_motion, last_read, last_read_elapsed, last_fast_motion = 0;
+    static bool         in_motion   = false;
+    bool                fast_motion = false;
+
+    pimoroni_data_t pimoroni_data = {0};
 
     if (error_count < PIMORONI_TRACKBALL_ERROR_COUNT) {
         i2c_status_t status = read_pimoroni_trackball(&pimoroni_data);
+        last_read_elapsed   = timer_elapsed_fast(last_read);
+        last_read           = timer_read_fast();
 
         if (status == I2C_STATUS_SUCCESS) {
             error_count = 0;
 
             if (!(pimoroni_data.click & 128)) {
                 mouse_report.buttons = pointing_device_handle_buttons(mouse_report.buttons, false, POINTING_DEVICE_BUTTON1);
                 if (!debounce) {
-                    x_offset += pimoroni_trackball_get_offsets(pimoroni_data.right, pimoroni_data.left, PIMORONI_TRACKBALL_SCALE);
-                    y_offset += pimoroni_trackball_get_offsets(pimoroni_data.down, pimoroni_data.up, PIMORONI_TRACKBALL_SCALE);
-                    mouse_report.x = pimoroni_trackball_adapt_values(&x_offset);
-                    mouse_report.y = pimoroni_trackball_adapt_values(&y_offset);
+                    if (pimoroni_data.left || pimoroni_data.right || pimoroni_data.up || pimoroni_data.down) {
+                        last_motion = last_read;
+                        if (!in_motion) {
+                            in_motion = true;
+                        }
+                    } else {
+                        if (timer_elapsed_fast(last_motion) > PIMORONI_TRACKBALL_RESET_DELAY) {
+                            in_motion = false;
+                            began_motion += last_read_elapsed * 4;
+                            if (began_motion > last_read) {
+                                began_motion = last_read;
+                            }
+                        }
+                    }
+
+                    if (in_motion) {
+                        mouse_report.x = (int8_t)pimoroni_data.left - (int8_t)pimoroni_data.right;
+                        mouse_report.y = (int8_t)pimoroni_data.up - (int8_t)pimoroni_data.down;
```

Comment:
x and y can be int16_t.  `mouse_xy_report_t` is typedef'd to the correct range, and may be more appropriate here. 
```suggestion
                        mouse_report.x = (mouse_xy_report_t)pimoroni_data.left - (mouse_xy_report_t)pimoroni_data.right;
                        mouse_report.y = (mouse_xy_report_t)pimoroni_data.up - (mouse_xy_report_t)pimoroni_data.down;
```

Reply:
Yeah, figured as much.